graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -0,0 +1,264 @@
+        location.query_path: location_info.type
+        for location, location_info in query_metadata_table.registered_locations
+    }
+    query_path_to_output_fields = assign_output_fields_to_nodes(construct_result, location_types)
+    return SqlQueryTree(tree_root, query_path_to_location_info, query_path_to_filter,
+                        query_path_to_output_fields, query_path_to_tag_fields)
+
+
+def get_tag_fields(expression):
+    """Return an iterator over the tag fields of an expression."""
+    if isinstance(expression, expressions.ContextField):
+        yield expression
+    if isinstance(expression, expressions.BinaryComposition):
+        for context_field in get_tag_fields(expression.left):
+            yield context_field
+        for context_field in get_tag_fields(expression.right):
+            yield context_field
+
+
+def assign_output_fields_to_nodes(construct_result, location_types):
+    """Assign the output fields of a ConstructResult block to their respective SqlNodes."""
+    query_path_to_output_fields = defaultdict(dict)
+    for field_alias, field in six.iteritems(construct_result.fields):
+        if field_alias in RESERVED_COLUMN_NAMES:
+            raise exceptions.GraphQLCompilationError(
+                u'Column name "{}" is reserved and cannot be used in output directive.'.format(
+                    field_alias))
+        field_name = field.location.field
+        if field_name in UNSUPPORTED_META_FIELDS:
+            raise exceptions.GraphQLNotSupportedByBackendError(
+                u'"{}" is unsupported for output.'.format(UNSUPPORTED_META_FIELDS[field_name]))
+        output_query_path = field.location.query_path
+        output_field_info = (field, location_types[output_query_path], False)
+        query_path_to_output_fields[output_query_path][field_alias] = output_field_info
+    return query_path_to_output_fields
+
+
+def get_block_index_to_location_map(ir_blocks):
+    """Associate each IR block with it's corresponding location, by index."""
+    block_to_location = {}
+    current_block_ixs = []
+    for num, ir_block in enumerate(ir_blocks):
+        current_block_ixs.append(num)
+        if isinstance(ir_block, blocks.MarkLocation):
+            for ix in current_block_ixs:
+                block_to_location[ix] = ir_block.location
+            current_block_ixs = []
+    return block_to_location
+
+
+def lower_construct_result(construct_result):
+    """Lower the ConstructResult IR Block."""
+    def visitor_fn(expression):
+        """Rewrite output fields in optional scopes to their if_true branch.
+
+        The SQL backend does not require special handling to get the correct semantics for such
+        output fields.
+        """
+        if not isinstance(expression, expressions.TernaryConditional):
+            return expression
+        return expression.if_true
+    return construct_result.visit_and_update_expressions(visitor_fn)
+
+
+def lower_unary_transformations(ir_blocks):
```

Comment:
Isn't `has_edge_degree` just a check against the cardinality of the join for a particular value?

Otherwise, I think this is probably fine for the time being. I think over time we'll want to move toward metadata-driven `NotSupportedByBackend` errors: the front-end generates IR and metadata, then the backend analyzes the metadata and complains if necessary.

Reply:
Yeah exactly, that's what it would be.

I've been thinking about this, and one approach would be to use window functions to do the counting, and then filter that on only rows with a certain count. So if the edge is from column `animal_id` to column `parentof_id`, we would have something like
`COUNT(*) OVER (PARTITION BY parentof_id) AS edge_degree`  and then in a second CTE filter out rows that have the correct target edge degree. 

With your second comment, do you mean something that identifies constructs used in the metadata (like `has_edge_degree` was used), and then using that to reject on the backend?

Reply:
That seems very reasonable, though exactly the sort of thing for which I'd hope to have a compiler in the loop -- I would not like to write that SQL query by hand :) We don't have to enable it in v1 of the SQL backend though, happy to save it for later.

Regarding the metadata, I'd like to extend the `QueryMetadataTable` to include information about all the locations and kinds of filters that were used (as well as lots of other info, like fold and output locations etc.). Then, before lowering or emitting SQL, the SQL backend could look over this metadata and decide whether anything in there is unsupported. This is how you'd also raise `NotSupportedByBackend` errors for `@fold` on anything but Postgres due to the lack of proper array aggregates -- the backend knows the target platform, and the frontend supplies the metadata for the backend to decide on.